File: app/src/main/java/edu/mills/heartoftaiwanese/MainActivity.java
package edu.mills.heartoftaiwanese;

import android.os.AsyncTask;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        final EditText chineseEditText = findViewById(R.id.editText);
        final EditText englishEditText = findViewById(R.id.editTextEng);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Button b = findViewById(R.id.submit_ch);
        b.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                findViewById(R.id.tw_result).setVisibility(View.GONE);
                findViewById(R.id.result).setVisibility((View.GONE));
                new ParseWordTask().execute((new LanguageContainer(chineseEditText
                        .getText().toString(), Word.LANGUAGE_CHINESE))
                );
            }
        });
        Button b2 = findViewById(R.id.submit_en);
        b2.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                findViewById(R.id.tw_result).setVisibility(View.GONE);
                findViewById(R.id.result).setVisibility((View.GONE));
                new ParseWordTask().execute((new LanguageContainer(englishEditText
                        .getText().toString(), Word.LANGUAGE_ENGLISH))
                );

            }
        });
        Button clearButton = findViewById(R.id.clear);
        clearButton.setOnClickListener(new View.OnClickListener() {
            /**
             * Called when a view has been clicked.
             *
             * @param v The view that was clicked.
             */
            @Override
            public void onClick(View v) {
                chineseEditText.setText("");
                englishEditText.setText("");
                findViewById(R.id.tw_result)
                        .setVisibility(View.GONE);
                findViewById(R.id.result)
                        .setVisibility(View.GONE);
            }
        });
    }

    private class LanguageContainer {
        private final int lang;
        private final String text;

        LanguageContainer(String input, int languageCode) {
            text = input;
            lang = languageCode;
        }

        int getLanguage() {
            return lang;
        }

        String getText() {
            return text;
        }
    }

    private class ParseWordTask extends AsyncTask<LanguageContainer, Integer, Boolean> {
        /**
         * Override this method to perform a computation on a background thread. The
         * specified parameters are the parameters passed to {@link #execute}
         * by the caller of this task.
         * <p>
         * This method can call {@link #publishProgress} to publish updates
         * on the UI thread.
         *
         * @see #onPreExecute()
         * @see #onPostExecute
         * @see #publishProgress
         */
        private Word w;

        @Override
        protected Boolean doInBackground(@org.jetbrains.annotations.NotNull LanguageContainer... lang) {
            w = new Word(lang[0].getText(), lang[0].getLanguage());
            return w.run();
        }

        @Override
        protected void onPostExecute(Boolean status) {
            if (status) {
                Log.d("MainActivity-Taiwanese", w.toString());
                findViewById(R.id.tw_result).setVisibility(View.VISIBLE);
                TextView t = findViewById(R.id.result);
                t.setVisibility(View.VISIBLE);
                t.setText(w.getTaiwanese());
            } else {
                String error = MainActivity.this.getText(R.string.error).toString();
                Toast.makeText(MainActivity.this, error, Toast.LENGTH_SHORT).show();
            }
            super.onPostExecute(status);
        }
    }
}

File: app/src/main/java/edu/mills/heartoftaiwanese/main/MainActivity.kt
package edu.mills.heartoftaiwanese.main

import android.os.AsyncTask
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import edu.mills.heartoftaiwanese.R
import edu.mills.heartoftaiwanese.WordRetriever
import edu.mills.heartoftaiwanese.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {
    companion object {
        const val LANGUAGE_ENGLISH = 1
        const val LANGUAGE_CHINESE = 0
        private const val kResultOk = "SUCCESS"
    }

    private lateinit var binding: ActivityMainBinding
    private lateinit var clView: View

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        initializeClickListeners()
    }

    private fun initializeClickListeners() {
        binding.submitCh.setOnClickListener {
            binding.twResult.visibility = View.GONE
            binding.result.visibility = View.GONE
            ParseWordTask().execute(
                LanguageContainer(
                    binding.editTextCh.text.toString(),
                    LANGUAGE_CHINESE
                )
            )
        }
        binding.submitEn.setOnClickListener {
            binding.twResult.visibility = View.GONE
            binding.result.visibility = View.GONE
            ParseWordTask().execute(
                LanguageContainer(
                    binding.editTextEng.text.toString(),
                    LANGUAGE_ENGLISH
                )
            )
        }
        binding.clearButton.setOnClickListener {
            binding.editTextCh.setText("")
            binding.editTextEng.setText("")
            binding.twResult.visibility = View.GONE
            binding.result.visibility = View.GONE
        }
    }

    private data class LanguageContainer internal constructor(
        internal val text: String,
        internal val language: Int
    )

    private inner class ParseWordTask : AsyncTask<LanguageContainer, Int, String>() {
        /**
         * Override this method to perform a computation on a background thread. The
         * specified parameters are the parameters passed to [.execute]
         * by the caller of this task.
         *
         *
         * This method can call [.publishProgress] to publish updates
         * on the UI thread.
         *
         * @see .onPreExecute
         * @see .onPostExecute
         *
         * @see .publishProgress
         */
        private lateinit var w: WordRetriever

        override fun doInBackground(vararg lang: LanguageContainer): String? {
            if (lang[0].language == LANGUAGE_ENGLISH) {
                val chineseString = w.fetchChinese(lang[0].text)
                if (chineseString == WordRetriever.kUnknownError) {
                    return WordRetriever.kUnknownError
                } else if (chineseString == WordRetriever.kRateLimited) {
                    return WordRetriever.kRateLimited
                } else {
                    w = WordRetriever(chineseString)
                }
//                Log.d("MainActivity", w.toString())
            } else {
                w = WordRetriever(lang[0].text)
            }
            val status = w.run()
            if (!status) {
                return WordRetriever.kUnknownError
            } else {
                return kResultOk
            }
        }

        override fun onPostExecute(status: String) {
            if (status == WordRetriever.kRateLimited) {
                Toast.makeText(
                    this@MainActivity,
                    getText(R.string.too_many_requests),
                    Toast.LENGTH_LONG
                ).show()
            } else if (status == WordRetriever.kUnknownError) {
                Toast.makeText(this@MainActivity, getText(R.string.error), Toast.LENGTH_LONG).show()
            } else {
                Log.d("MainActivity-Taiwanese", w.toString())
                binding.twResult.visibility = View.VISIBLE
                binding.result.visibility = View.VISIBLE
                binding.result.text = w.taiwanese
            }
            super.onPostExecute(status)
        }
    }
}
